plenoptic/synthesize/geodesic.py

```diff
@@ -1,155 +1,304 @@
+            set the random number generator
+        """
+        self.lmbda = lmbda
 
-        for i in range(max_iter):
-            loss = self._optimizer_step(i, max_iter)
+        torch.manual_seed(seed)
+        if optimizer == 'adam':
+            self.optimizer = optim.Adam([self.x],
+                                        lr=learning_rate, amsgrad=True)
+        elif optimizer == 'sgd':
+            self.optimizer = optim.SGD([self.x],
+                                       lr=learning_rate, momentum=0.9)
+        elif isinstance(optimizer, optim.Optimizer):
+            self.optimizer = optimizer
+
+        pbar = tqdm(range(max_iter))
+        for i in pbar:
+            loss = self._optimizer_step(i, pbar)
+
+            # storing some information
             self.loss.append(loss.item())
-
-        self.gamma.copy_(self.img.data.squeeze()).add_(0.5).mul_(255)
-
-        return self.gamma
+            self.geodesic = torch.cat((self.xA, self.x.data, self.xB), 0)
+            # TODO flag to store progress or not
+            self.dist_from_line.append(distance_from_line(
+                                        self.geodesic))
+
+            if loss.item() < 1e-6:
+                raise Exception("""the geodesic matches the representation
+                                straight line up to floating point
+                                precision""")
+
+    def plot_loss(self):
+        plt.semilogy(self.loss)
+        plt.xlabel('iter step')
+        plt.ylabel('loss value')
+        plt.show()
+
+    def plot_distance_from_line(self, vid=None, iteration=None,
+                                figsize=(7, 5)):
+        """visual diagnostic of geodesic linearity in representation space.
+
+        Parameters
+        ----------
+        vid : torch.Tensor, optional
+            natural video that bridges the anchor points
+        iteration : int, optional
+            plot the geodesic at a given step number of the optimization
+        figsize : tuple, optional
+            set the dimension of the figure
+
+        Returns
+        -------
+        fig: matplotilb figure object
+        """
+        fig, ax = plt.subplots(1, 1, figsize=figsize)
+        ax.plot(to_numpy(distance_from_line(self.pixelfade)),
+                'g-o', label='pixelfade')
+
+        if iteration is None:
+            distance = distance_from_line(self.geodesic)
+        else:
+            distance = self.dist_from_line[iteration]
+        ax.plot(to_numpy(distance), 'r-o', label='geodesic')
+
+        if vid is not None:
+            ax.plot(to_numpy(distance_from_line(vid)),
+                    'b-o', label='video')
+        ax.set(xlabel='projection on representation line',
+               ylabel='distance from representation line')
+        ax.legend(loc=1)
+
+        return fig
+
+    def animate_distance_from_line(self, vid=None, framerate=25):
+        """dynamic visualisation of geodesic linearity along the optimization process
+
```

Comment:
I think the description should mention that this basically animates `plot_distance_from_line` over iterations of synthesis, showing how the distance of the geodesic changed. That's not clear to me from this description.